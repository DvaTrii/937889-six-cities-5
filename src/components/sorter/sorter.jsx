import React from "react";
import {connect} from "react-redux";
import {compose} from "redux";
import PropTypes from "prop-types";

import {withActiveFlag} from "../hocs/withActiveFlag/withActiveFlag";
import {SorterType, SorterActiveClass, SorterListOpenedClass} from "../../const";
import {setActiveSorter} from "../../store/app/actions";


const Sorter = ({isActive, onActiveChange, activeSorter, setActiveSorterAction}) => {
  return (
    <form className="places__sorting" action="#" method="get">
      <span className="places__sorting-caption">Sort by</span>
      <span className="places__sorting-type"
        tabIndex="0"
        onClick={onActiveChange}>
        {activeSorter}
        <svg className="places__sorting-arrow" width="7" height="4">
          <use xlinkHref="#icon-arrow-select"></use>
        </svg>
      </span>
      <ul className={`places__options places__options--custom
        ${isActive ?
      SorterListOpenedClass.OPENED : ``}`}>

        {
          Object.values(SorterType).map((it) => (
            <li key={it}
              className={`places__option ${activeSorter === it ? SorterActiveClass.ACTIVE : ``}`}
              onClick={() => {
                setActiveSorterAction(it);
                onActiveChange();
              }}
              tabIndex="0">
              {it}
            </li>))
        }

      </ul>
      {/* <select className="places__sorting-type" id="places-sorting">*/}
      {/*  <option className="places__option" value="popular" selected="">Popular</option>*/}
      {/*  <option className="places__option" value="to-high">Price: low to high</option>*/}
      {/*  <option className="places__option" value="to-low">Price: high to low</option>*/}
      {/*  <option className="places__option" value="top-rated">Top rated first</option>*/}
      {/* </select>*/}
    </form>
  );
};

Sorter.propTypes = {
  isActive: PropTypes.bool.isRequired,
  onActiveChange: PropTypes.func.isRequired,
  activeSorter: PropTypes.string.isRequired,
  setActiveSorterAction: PropTypes.func.isRequired
};

const mapStateToProps = ({APP}) => {
  return ({
    activeSorter: APP.activeSorter,
  });
};

const mapDispatchToProps = (dispatch) => ({
  setActiveSorterAction(sorter) {
    dispatch(setActiveSorter(sorter));
  }
});

export {Sorter};
export default compose(withActiveFlag, connect(mapStateToProps, mapDispatchToProps))(Sorter);
