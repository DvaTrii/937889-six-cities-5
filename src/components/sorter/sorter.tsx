import * as React from "react";
import {connect} from "react-redux";

import {SorterType, SorterActiveClass, SorterListOpenedClass} from "../../const";
import {getActiveSorter, getActiveCity} from "../../store/app/selectors";
import {setActiveSorter} from "../../store/app/actions";

type Props = {
  activeSorter: string,
  setActiveSorterAction: (string) => void,
  activeCity: string
}

const Sorter: React.FC<Props> = ({activeSorter, setActiveSorterAction, activeCity}) => {

  const [isActive, setIsActive] = React.useState<boolean>(false);
  const toggleIsActive = () => setIsActive(!isActive);

  React.useEffect(() => {
    setIsActive(false);
  }, [activeCity]);

  return (
    <form className="places__sorting" action="#" method="get">
      <span className="places__sorting-caption">Sort by</span>
      <span className="places__sorting-type"
        // tabIndex="0"
        onClick={toggleIsActive}>
        {activeSorter}
        <svg className="places__sorting-arrow" width="7" height="4">
          <use xlinkHref="#icon-arrow-select"/>
        </svg>
      </span>
      <ul className={`places__options places__options--custom
        ${isActive ?
      SorterListOpenedClass.OPENED : ``}`}>

        {
          Object.values(SorterType).map((it: string) => (
            <li key={it}
              className={`places__option ${activeSorter === it ? SorterActiveClass.ACTIVE : ``}`}
              onClick={() => {
                setActiveSorterAction(it);
                toggleIsActive();
              }}
              // tabIndex="0"
              >
              {it}
            </li>))
        }

      </ul>
    </form>
  );
};

const mapStateToProps = (state) => {
  return ({
    activeSorter: getActiveSorter(state),
    activeCity: getActiveCity(state)
  });
};

const mapDispatchToProps = (dispatch) => ({
  setActiveSorterAction(sorter) {
    dispatch(setActiveSorter(sorter));
  }
});

export {Sorter};
export default connect(mapStateToProps, mapDispatchToProps)(Sorter);
