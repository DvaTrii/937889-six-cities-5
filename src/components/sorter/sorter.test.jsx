import React from "react";
import renderer from "react-test-renderer";
import {Sorter} from "./sorter";

it(`Should Sorter render properly`, () => {

  const noop = () => {};
  const activeSorter = `Popular`;
  const activeCity = `Paris`;

  const tree = renderer
    .create(
        <Sorter
          activeSorter={activeSorter}
          setActiveSorterAction={noop}
          activeCity={activeCity}
        />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
