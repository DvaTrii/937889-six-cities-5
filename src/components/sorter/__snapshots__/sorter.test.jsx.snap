// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Sorter> render Should Sorter with activeFlag = false render properly 1`] = `
<form
  action="#"
  className="places__sorting"
  method="get"
>
  <span
    className="places__sorting-caption"
  >
    Sort by
  </span>
  <span
    className="places__sorting-type"
    onClick={[Function]}
    tabIndex="0"
  >
    Popular
    <svg
      className="places__sorting-arrow"
      height="4"
      width="7"
    >
      <use
        xlinkHref="#icon-arrow-select"
      />
    </svg>
  </span>
  <ul
    className="places__options places__options--custom
        "
  >
    <li
      className="places__option places__option--active"
      onClick={[Function]}
      tabIndex="0"
    >
      Popular
    </li>
    <li
      className="places__option "
      onClick={[Function]}
      tabIndex="0"
    >
      Price: low to high
    </li>
    <li
      className="places__option "
      onClick={[Function]}
      tabIndex="0"
    >
      Price: high to low
    </li>
    <li
      className="places__option "
      onClick={[Function]}
      tabIndex="0"
    >
      Top rated first
    </li>
  </ul>
</form>
`;

exports[`<Sorter> render Should Sorter with activeFlag = true render properly 1`] = `
<form
  action="#"
  className="places__sorting"
  method="get"
>
  <span
    className="places__sorting-caption"
  >
    Sort by
  </span>
  <span
    className="places__sorting-type"
    onClick={[Function]}
    tabIndex="0"
  >
    Popular
    <svg
      className="places__sorting-arrow"
      height="4"
      width="7"
    >
      <use
        xlinkHref="#icon-arrow-select"
      />
    </svg>
  </span>
  <ul
    className="places__options places__options--custom
        places__options--opened"
  >
    <li
      className="places__option places__option--active"
      onClick={[Function]}
      tabIndex="0"
    >
      Popular
    </li>
    <li
      className="places__option "
      onClick={[Function]}
      tabIndex="0"
    >
      Price: low to high
    </li>
    <li
      className="places__option "
      onClick={[Function]}
      tabIndex="0"
    >
      Price: high to low
    </li>
    <li
      className="places__option "
      onClick={[Function]}
      tabIndex="0"
    >
      Top rated first
    </li>
  </ul>
</form>
`;
